cuda/solver/lower_trs_kernels.cu

```diff
@@ -62,6 +64,173 @@ namespace cuda {
+                      thrust::complex<ValueType> value)
+{
+    auto real = reinterpret_cast<ValueType *>(values);
+    auto imag = real + 1;
+    store(real, 2 * index, value.real());
+    store(imag, 2 * index, value.imag());
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ void sptrsv_naive_caching_kernel(
+    const IndexType *const rowptrs, const IndexType *const colidxs,
+    const ValueType *const vals, const ValueType *const b, ValueType *const x,
+    volatile uint32 *const ready, const size_type n)
+{
+    const auto gid = thread::get_thread_id_flat<IndexType>();
+    const auto row = gid;
+
+    if (row >= n) {
+        return;
+    }
+
+    const auto self_shmem_id = gid / default_block_size;
+    const auto self_shid = gid % default_block_size;
+
+    volatile __shared__ uint32 ready_s[default_block_size];
+    __shared__ UninitializedArray<ValueType, default_block_size> x_s_array;
+    ValueType *x_s = x_s_array;
+    ready_s[self_shid] = 0;
+    x_s[self_shid] = zero<ValueType>();
+
+    __syncthreads();
+
+    const auto row_start = rowptrs[row];
+    const auto row_end = rowptrs[row + 1] - 1;
+
+    ValueType sum = 0.0;
+    for (IndexType i = row_start; i < row_end; ++i) {
+        const auto dependency = colidxs[i];
+        volatile auto is_ready = &ready[dependency];
+        auto x_p = &x[dependency];
+
+        const bool shmem_possible =
+            dependency / default_block_size == self_shmem_id;
+        if (shmem_possible) {
+            const auto dependency_shid = dependency % default_block_size;
+            is_ready = &ready_s[dependency_shid];
+            x_p = &x_s[dependency_shid];
+        }
+
+        uint32 is_ready_v = false;
+        while (!is_ready_v) {
+            is_ready_v = *is_ready;
+        }
+
+        sum += load(x_p, 0) * vals[i];
+    }
+
+    const auto r = (b[row] - sum) / vals[row_end];
+
+    store(x_s, self_shid, r);
+    __threadfence_block();
+    ready_s[self_shid] = 1;
+
+    x[row] = r;
+    __threadfence();
+    ready[row] = 1;
+}
+
+
+template <typename ValueType, typename IndexType>
+__global__ void sptrsv_naive_legacy_kernel(
+    const IndexType *const rowptrs, const IndexType *const colidxs,
+    const ValueType *const vals, const ValueType *const b, ValueType *const x,
+    volatile uint32 *const ready, const size_type n)
+{
+    const auto gid = thread::get_thread_id_flat<IndexType>();
+    const auto row = gid;
+
+    if (row >= n) {
+        return;
+    }
+
+    const auto row_end = rowptrs[row + 1] - 1;
+
+    ValueType sum = 0.0;
+    auto j = rowptrs[row];
+    while (j <= row_end) {
+        auto col = colidxs[j];
+        while (ready[col]) {
+            sum += vals[j] * load(x, col);
+            ++j;
+            col = colidxs[j];
+        }
+        if (row == col) {
+            store(x, row, (b[row] - sum) / vals[row_end]);
+            ++j;
+            __threadfence();
```

Comment:
Does it matter whether the threadfence is predicated?

Reply:
I did a quick test on one matrix, there performance indeed improves by ~5% by factoring it out. Good catch!

Reply:
I have to correct this: I by accident noticed that, at least for one matrix, the fallback version demands the minimal block size of 32 threads for acceptable performance. In that scenario, the threadfence should be predicated. Then again, these measurements are from my post-volta GPU, perhaps it looks different on a P100.